windows-agent/internal/proservices/landscape/executor.go

```diff
@@ -189,3 +196,119 @@ func (e executor) setDefault(ctx context.Context, cmd *landscapeapi.Command_SetD
 func (e executor) shutdownHost(ctx context.Context, cmd *landscapeapi.Command_ShutdownHost) error {
 	return gowsl.Shutdown(ctx)
 }
+
+func installFromMicrosoftStore(ctx context.Context, distro gowsl.Distro) error {
+	if err := gowsl.Install(ctx, distro.Name()); err != nil {
+		return err
+	}
+
+	if err := distroinstall.InstallFromExecutable(ctx, distro); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func installFromURL(ctx context.Context, homeDir string, distro gowsl.Distro, rootfs *landscapeapi.Command_Install_Rootfs) (err error) {
+	// Since we are going to remove the tarball soon, I'd say it's fine to write it to a subfolder of the current working directory,
+	// as it will be somewhere inside the Windows Apps private folders:
+	// %USERPROFILE%\AppData\Local\Packages\CanonicalGroupLimited.UbuntuPro_79rhkp1fndgsc\LocalState\
+	cwd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	tmpDir := filepath.Join(cwd, distro.Name())
+	if err := os.MkdirAll(tmpDir, 0700); err != nil {
+		return err
+	}
+	// Remove tarball once installed
+	defer os.RemoveAll(tmpDir)
+
+	tarball := filepath.Join(tmpDir, distro.Name()+".tar.gz")
+
+	f, err := os.Create(tarball)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	if err := download(f, rootfs.GetUrl()); err != nil {
+		return err
+	}
+
+	if err := checksumMatches(tarball, rootfs.GetSha256Sum()); err != nil {
+		return err
+	}
+
+	// Create the directory that will contain the vhdx
+	vhdxDir := filepath.Join(homeDir, "WSL", distro.Name())
+	if err := os.MkdirAll(vhdxDir, 0700); err != nil {
+		return err
+	}
+
+	if _, err := gowsl.Import(ctx, distro.Name(), tarball, vhdxDir); err != nil {
+		_ = os.RemoveAll(vhdxDir)
+		return err
+	}
+	return nil
+}
+
+func download(w io.Writer, url string) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("could not download %q: %v", url, err)
+		}
+	}()
+
+	//nolint:gosec // G107 expects url to be const, but we are reading it from Landscape.
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 400 {
+		return fmt.Errorf("http request failed with code %d", resp.StatusCode)
+	}
+
+	if _, err := io.Copy(w, resp.Body); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checksumMatches(path, wantChecksum string) (err error) {
```

Comment:
I would return 2 values here: (bool, err)

bool is the returned value if the checksum matches, and error only in the internal (like reading the file)

Also, I would instead of taking a path, use an io.Reader here, but see check my suggestion before :)